src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -80,11 +81,37 @@ public CursorsService(final ZooKeeperHolder zkHolder,
 
         validateStreamId(cursors, streamId, subscriptionId);
 
+        LOG.debug("[COMMIT_CURSORS] stream IDs validation finished");
+
+        final Iterator<EventTypePartition> distinctPartitions = cursors.stream()
```

Comment:
Okay, got it! 
It looks like you go through the whole list of cursors every time. It makes more sense then to pass the preprocessed map to next method calls.
cursors.stream().collect(Collectors.groupingBy((s -> s.getEventType() + "_" + s.getPartition())))
In that case you will just grab the list for particular event type partition and work with it: compare offsets and commit the largest one. It seems legitimate since the order should be correct per partition.